File: ios/Classes/CBManager.swift
//
//  CBManager.swift
//  Runner
//
//  Created by Luca Christille on 14/08/18.
//  Copyright © 2018 The Chromium Authors. All rights reserved.
//

import Foundation
import CouchbaseLiteSwift

class CBManager {
    static let instance = CBManager();
    private var mDatabase : Dictionary<String, Database> = Dictionary();
    private var mReplConfig : ReplicatorConfiguration! = nil;
    private var mReplicator : Replicator! = nil;
    private var defaultDatabase = "defaultDatabase";
    
    private init() {}
    
    func getDatabase() -> Database? {
        if let result = mDatabase[defaultDatabase] {
            return result;
        } else {
            return nil;
        }
    }
    
    func getDatabase(name : String) -> Database? {
        if let result = mDatabase[name] {
            return result;
        } else {
            return nil;
        }
    }
    
    func saveDocument(map: Dictionary<String, Any>) throws -> String? {
        let mutableDocument: MutableDocument = MutableDocument(data: map);
        try mDatabase[defaultDatabase]?.saveDocument(mutableDocument)
        return mutableDocument.id;
    }
    
    func saveDocumentWithId(id : String, map: Dictionary<String, Any>) throws -> String? {
        let mutableDocument: MutableDocument = MutableDocument(id: id, data: map)
        try mDatabase[defaultDatabase]?.saveDocument(mutableDocument)
        return mutableDocument.id
    }
    
    func getDocumentWithId(id : String) -> NSDictionary? {
        let resultMap: NSMutableDictionary = NSMutableDictionary.init()
        if let defaultDb: Database = getDatabase() {
            if let document: Document = defaultDb.document(withID: id) {
                let retrievedDocument: NSDictionary = NSDictionary.init(dictionary: document.toDictionary())
                // It is a repetition due to implementation of Document Dart Class
                resultMap["id"] = id
                resultMap["doc"] = retrievedDocument
            } else {
                resultMap["id"] = id
                resultMap["doc"] = NSDictionary.init()
            }
        }
        return NSDictionary.init(dictionary: resultMap)
    }

    func getDocumentsWith(key : String, value: String) -> NSDictionary? {
        let resultMap: NSMutableDictionary = NSMutableDictionary();
        if let defaultDb: Database = getDatabase() {
            let query = QueryBuilder
                .select(SelectResult.all(), SelectResult.expression(Meta.id))
                .from(DataSource.database(defaultDb))
                .where(Expression.property(key).equalTo(Expression.string(value)))
            do {
                let result = try query.execute()
                let docs = result.allResults().map{(result)->NSDictionary in
                    let ret = NSMutableDictionary();
                    if let doc = result.dictionary(forKey: defaultDb.name ?? defaultDatabase) {
                        ret["doc"] = doc.toDictionary();
                        let id = result.string(forKey: "id");
                        ret["id"] = id
                    }
                    return ret;
                    };
                resultMap["docs"] = docs;
            } catch {
                resultMap["docs"] = nil;
            }
        }
        return resultMap;
    }


    func getAllDocuments() -> NSDictionary {
        let resultMap: NSMutableDictionary = NSMutableDictionary();
        if let defaultDb: Database = getDatabase() {
            let query = QueryBuilder
                .select(SelectResult.all(), SelectResult.expression(Meta.id))
                .from(DataSource.database(defaultDb))
            do {
                let result = try query.execute()
                let docs = result.allResults().map{(result)->NSDictionary in
                    let ret = NSMutableDictionary();
                    if let doc = result.dictionary(forKey: defaultDb.name ?? defaultDatabase) {
                        ret["doc"] = doc.toDictionary();
                        let id = result.string(forKey: "id");
                        ret["id"] = id
                    }
                    return ret;
                    };
                resultMap["docs"] = docs;
            } catch {
                resultMap["docs"] = nil;
            }
        }
        return resultMap;
    }
    
    func purgeDocument(docId: String) -> Bool {
        if let defaultDb: Database = getDatabase() {
            if let doc: Document = defaultDb.document(withID: docId) {
                try! defaultDb.purgeDocument(doc);
                return true;
            }
        }
        return false;
    }

    func initDatabaseWithName(name: String){
        if mDatabase.keys.contains(name) {
            defaultDatabase = name
        } else {
            do {
                let newDatabase = try Database(name: name)
                // Database.setLogLevel(level: LogLevel.verbose, domain: LogDomain.replicator)
                mDatabase[name] = newDatabase
                defaultDatabase = name
            } catch {
                print("Error initializing new database")
            }
        }
    }
    
    func setReplicatorEndpoint(endpoint: String) {
        let targetEndpoint = URLEndpoint(url: URL(string: endpoint)!)
        mReplConfig = ReplicatorConfiguration(database: getDatabase()!, target: targetEndpoint)
    }
    
    func setReplicatorType(type: String) -> String {
        var settedType: ReplicatorType = ReplicatorType.pull
        if (type == "PUSH") {
            settedType = .push
        } else if (type == "PULL") {
            settedType = .pull
        } else if (type == "PUSH_AND_PULL") {
            settedType = .pushAndPull
        }
        mReplConfig?.replicatorType = settedType
        switch(mReplConfig?.replicatorType.rawValue) {
        case (0):
            return "PUSH_AND_PULL"
        case (1):
            return "PUSH"
        case(2):
            return "PULL"
        default:
            return ""
        }
    }
    
    func setReplicatorAuthentication(auth: [String:String]) -> String {
        if let username = auth["username"], let password = auth["password"] {
            mReplConfig?.authenticator = BasicAuthenticator(username: username, password: password)
        }
        return mReplConfig.authenticator.debugDescription
    }
    
    func setReplicatorSessionAuthentication(sessionID: String?) {
        if ((sessionID) != nil) {
            if ((mReplConfig) != nil) {
            mReplConfig.authenticator = SessionAuthenticator(sessionID: sessionID!)
            }
        }
    }
    
    func setReplicatorContinuous(isContinuous: Bool) -> Bool {
        if ((mReplConfig) != nil) {
            mReplConfig?.continuous = isContinuous
            return mReplConfig!.continuous
        }
        return false
    }
    
    func initReplicator() {
        mReplicator = Replicator(config: mReplConfig)
    }
    
    func startReplication() {
        if ((mReplicator) != nil) {
            mReplicator.start()
        }
    }
    
    func stopReplication() {
        if ((mReplicator) != nil) {
            mReplicator.stop()
            mReplicator = nil
        }
    }
    
    func getReplicator() -> Replicator {
        return mReplicator
    }
}

File: android/src/main/java/it/oltrenuovefrontiere/fluttercouch/CBManager.java
package it.oltrenuovefrontiere.fluttercouch;

import android.content.Context;
import android.util.Log;

import com.couchbase.lite.BasicAuthenticator;
import com.couchbase.lite.CouchbaseLiteException;
import com.couchbase.lite.DataSource;
import com.couchbase.lite.Database;
import com.couchbase.lite.DatabaseConfiguration;
import com.couchbase.lite.Document;
import com.couchbase.lite.Endpoint;
import com.couchbase.lite.Expression;
import com.couchbase.lite.Meta;
import com.couchbase.lite.MutableDocument;
import com.couchbase.lite.Query;
import com.couchbase.lite.QueryBuilder;
import com.couchbase.lite.Replicator;
import com.couchbase.lite.ReplicatorConfiguration;
import com.couchbase.lite.Result;
import com.couchbase.lite.ResultSet;
import com.couchbase.lite.SelectResult;
import com.couchbase.lite.SessionAuthenticator;
import com.couchbase.lite.URLEndpoint;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class CBManager {

    private HashMap<String, Database> mDatabase = new HashMap<>();
    private ReplicatorConfiguration mReplConfig;
    private Replicator mReplicator;
    private String defaultDatabase = "defaultDatabase";

    public CBManager() {
    }

    public Database getDatabase() {
        return mDatabase.get(defaultDatabase);
    }

    public Database getDatabase(String name) {
        if (mDatabase.containsKey(name)) {
            return mDatabase.get(name);
        }
        return null;
    }

    public String saveDocument(Map<String, Object> _map) throws CouchbaseLiteException {
        MutableDocument mutableDoc = new MutableDocument(_map);
        mDatabase.get(defaultDatabase).save(mutableDoc);
        return mutableDoc.getId();
    }

    public String saveDocumentWithId(String _id, Map<String, Object> _map) throws CouchbaseLiteException {
        MutableDocument mutableDoc = new MutableDocument(_id, _map);
        mDatabase.get(defaultDatabase).save(mutableDoc);
        return mutableDoc.getId();
    }

    public Map<String, Object> getDocumentWithId(String _id) throws CouchbaseLiteException {
        Database defaultDb = getDatabase();
        HashMap<String, Object> resultMap = new HashMap<String, Object>();
        if (defaultDatabase != null) {
            try {
                Document document = defaultDb.getDocument(_id);
                if (document != null) {
                    resultMap.put("doc", document.toMap());
                    resultMap.put("id", _id);
                } else {
                    resultMap.put("doc", null);
                    resultMap.put("id", _id);
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
        return resultMap;
    }

    public Map<String, Object> getDocumentsWith(String key, String value) throws CouchbaseLiteException {
        Database defaultDb = getDatabase();
        HashMap<String, Object> resultMap = new HashMap<String, Object>();
        Query query = QueryBuilder.select(SelectResult.all(), SelectResult.expression(Meta.id))
                .from(DataSource.database(defaultDb))
                .where(Expression.property(key).equalTo(Expression.string(value)));
        try {
            ResultSet result = query.execute();
            ArrayList docs = new ArrayList();
            String dbName = defaultDb.getName();
            for (Result res : result.allResults()) {
                HashMap<String, Object> ret = new HashMap<String, Object>();
                Object doc = res.toMap().get(dbName);
                ret.put("doc", doc);
                ret.put("id", res.getString("id"));
                docs.add(ret);
            }
            resultMap.put("docs", docs);
        } catch (Exception e) {
            e.printStackTrace();
            resultMap.put("docs", null);
        }
        return resultMap;
    }

    public Map<String, Object> getAllDocuments() throws CouchbaseLiteException {
        Database defaultDb = getDatabase();
        HashMap<String, Object> resultMap = new HashMap<String, Object>();
        Query query = QueryBuilder.select(SelectResult.all(), SelectResult.expression(Meta.id))
                .from(DataSource.database(defaultDb));
        try {
            ResultSet result = query.execute();
            ArrayList docs = new ArrayList();
            String dbName = defaultDb.getName();
            for (Result res : result.allResults()) {
                HashMap<String, Object> ret = new HashMap<String, Object>();
                Object doc = res.toMap().get(dbName);
                ret.put("doc", doc);
                ret.put("id", res.getString("id"));
                docs.add(ret);
            }
            resultMap.put("docs", docs);
        } catch (Exception e) {
            e.printStackTrace();
            resultMap.put("docs", null);
        }
        return resultMap;
    }

    public void purgeDocument(String _id) throws CouchbaseLiteException {
        Database defaultDb = getDatabase();
        Document document = defaultDb.getDocument(_id);
        if (document != null) {
            defaultDb.purge(document);
        }
    }

    public void initDatabaseWithName(String _name) throws CouchbaseLiteException {
        DatabaseConfiguration config = new DatabaseConfiguration(FluttercouchPlugin.context);
        if (!mDatabase.containsKey(_name)) {
            defaultDatabase = _name;
            // Database.setLogLevel(LogDomain.REPLICATOR, LogLevel.VERBOSE);
            mDatabase.put(_name, new Database(_name, config));
        }
    }

    public String setReplicatorEndpoint(String _endpoint) throws URISyntaxException {
        Endpoint targetEndpoint = new URLEndpoint(new URI(_endpoint));
        mReplConfig = new ReplicatorConfiguration(mDatabase.get(defaultDatabase), targetEndpoint);
        return mReplConfig.getTarget().toString();
    }

    public String setReplicatorType(String _type) throws CouchbaseLiteException {
        ReplicatorConfiguration.ReplicatorType settedType = ReplicatorConfiguration.ReplicatorType.PULL;
        if (_type.equals("PUSH")) {
            settedType = ReplicatorConfiguration.ReplicatorType.PUSH;
        } else if (_type.equals("PULL")) {
            settedType = ReplicatorConfiguration.ReplicatorType.PULL;
        } else if (_type.equals("PUSH_AND_PULL")) {
            settedType = ReplicatorConfiguration.ReplicatorType.PUSH_AND_PULL;
        }
        mReplConfig.setReplicatorType(settedType);
        return settedType.toString();
    }

    public String setReplicatorBasicAuthentication(Map<String, String> _auth) throws Exception {
        if (_auth.containsKey("username") && _auth.containsKey("password")) {
            mReplConfig.setAuthenticator(new BasicAuthenticator(_auth.get("username"), _auth.get("password")));
        } else {
            throw new Exception();
        }
        return mReplConfig.getAuthenticator().toString();
    }

    public String setReplicatorSessionAuthentication(String sessionID) throws Exception {
        if (sessionID != null) {
            mReplConfig.setAuthenticator(new SessionAuthenticator(sessionID));
        } else {
            throw new Exception();
        }
        return mReplConfig.getAuthenticator().toString();
    }

    public boolean setReplicatorContinuous(boolean _continuous) {
        mReplConfig.setContinuous(_continuous);
        return mReplConfig.isContinuous();
    }

    public void initReplicator() {
        mReplicator = new Replicator(mReplConfig);
    }

    public void startReplicator() {
        mReplicator.start();
    }

    public void stopReplicator() {
        mReplicator.stop();
        mReplicator = null;
    }

    public Replicator getReplicator() {
        return mReplicator;
    }
}
